### PR TITLE
Asynchronous support via checkpoints for @BeforeEach and @BeforeAll

### LINKED ISSUE #22: Asynchronous support via checkpoints for @BeforeEach and @BeforeAll

### PR DESCRIPTION
**Pull request by vietj**
fixes #22

### CODE REVIEW

**Comment by vietj** [comment]
will do

**Comment by vietj** [comment]
Documentation added

File: src/test/java/io/vertx/junit5/VertxExtensionTest.java [hunk 1]
```diff
@@ -17,7 +17,8 @@
 package io.vertx.junit5;
 
 import io.vertx.core.Vertx;
-import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.BeforeAll;
```
Comment (dadoonet) [inline src/test/java/io/vertx/junit5/VertxExtensionTest.java]: I don't think that those imports are needed, but this may be a side effect of #26

**Comment by jponge** [comment]
So under JDK9 surefire executes the test that is tagged and that shall be excluded. 30 tests.

JDK8 is fine with 29 tests.

**Comment by jponge** [comment]
Fascinating: https://github.com/junit-team/junit5/issues/1188

**Comment by jponge** [comment]
We're back to green on this branch @dadoonet thanks to your reports + missing test cases that @vietj reported and that exposed obvious bugs.

Can you please have a look?

File: .travis.yml [hunk 2]
```diff
@@ -3,10 +3,10 @@ addons:
   apt:
     packages:
       - oracle-java8-installer
-      - oracle-java9-installer
+#      - oracle-java9-installer
```
Comment (dadoonet) [inline .travis.yml]: Is it intentional then?

File: .travis.yml [hunk 3]
```diff
@@ -3,10 +3,10 @@ addons:
...
 
 language: java
 
 jdk:
   - oraclejdk8
-  - oraclejdk9
```
Comment (dadoonet) [inline .travis.yml]: Same here?

File: src/main/java/io/vertx/junit5/VertxExtension.java [hunk 4]
```diff
@@ -46,12 +53,48 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
...
+        timeoutUnit = annotation.timeUnit();
+      } else if (extensionContext.getRequiredTestClass().isAnnotationPresent(Timeout.class)) {
+        Timeout annotation = extensionContext.getRequiredTestClass().getAnnotation(Timeout.class);
+        timeoutDuration = annotation.value();
+        timeoutUnit = annotation.timeUnit();
+      }
+      if (context.awaitCompletion(timeoutDuration, timeoutUnit)) {
+        if (context.failed()) {
+          Throwable throwable = context.causeOfFailure();
+          if (throwable instanceof Exception) {
```
Comment (dadoonet) [inline src/main/java/io/vertx/junit5/VertxExtension.java]: I wonder if this should be done the other way like:

```java
if (throwable instanceof AssertionError) {
   throw (AssertionError) throwable;
} else {
   throw new AssertionError(throwable);
}
```

And it looks strange to me but I might be reading it wrong... :) 

Just wondering if someone throws a Throwable or any Throwable subclass (not an Exception or any Exception subclass).

File: .travis.yml (hunk 2 above)
Comment (dadoonet) [reply .travis.yml]: Forget it. I saw it is. May be add a comment in the yml file with a link to the issue?

Comment (jponge) [reply .travis.yml]: On the side I'm tracking the upstream issue, so as soon as it's been fixed I'll re-enable JDK9.

File: src/main/java/io/vertx/junit5/VertxExtension.java (hunk 4 above)
Comment (jponge) [reply src/main/java/io/vertx/junit5/VertxExtension.java]: Well the thing is that the JUnit5 extension API forces the method to have a `throws Exception` declaration, so we can't _simply_ throw the exception as-it, hence we have to wrap unchecked exceptions into checked exceptions.

Makes sense? (or am I wrong?)

**Comment by jponge** [comment]
Thanks @dadoonet. Have you checked with your own tests if that branch works?

**Comment by dadoonet** [comment]
> Have you checked with your own tests if that branch works?

Not yet... Give me some time to do it. :)

File: src/main/java/io/vertx/junit5/VertxExtension.java (hunk 4 above)
Comment (vietj) [reply src/main/java/io/vertx/junit5/VertxExtension.java]: why unchecked does not work ? I'm affraid it translates a JUnit failure into an assertion error and lose contextual information

**Comment by dadoonet** [comment]
Tests are failing on my laptop:

```sh
mvn clean install
```

```
[INFO] --- maven-surefire-plugin:2.19.1:test (default-test) @ vertx-junit5 ---
Downloading from central: https://repo.maven.apache.org/maven2/org/junit/platform/junit-platform-surefire-provider/1.0.2/junit-platform-surefire-provider-1.0.2.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/junit/platform/junit-platform-surefire-provider/1.0.2/junit-platform-surefire-provider-1.0.2.pom (2.7 kB at 2.3 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/junit/platform/junit-platform-surefire-provider/1.0.2/junit-platform-surefire-provider-1.0.2.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/junit/platform/junit-platform-surefire-provider/1.0.2/junit-platform-surefire-provider-1.0.2.jar (15 kB at 62 kB/s)

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running io.vertx.junit5.VertxTestContextTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.943 sec - in io.vertx.junit5.VertxTestContextTest
Running io.vertx.junit5.AsyncBeforeAllTest
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by io.netty.util.internal.ReflectionUtil (file:/Users/dpilato/.m2/repository/io/netty/netty-common/4.1.19.Final/netty-common-4.1.19.Final.jar) to constructor java.nio.DirectByteBuffer(long,int)
WARNING: Please consider reporting this to the maintainers of io.netty.util.internal.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.051 sec - in io.vertx.junit5.AsyncBeforeAllTest
Running io.vertx.junit5.AsyncBeforeCombinedTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.48 sec - in io.vertx.junit5.AsyncBeforeCombinedTest
Running io.vertx.junit5.VertxExtensionTest
Tests run: 9, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 0.297 sec <<< FAILURE! - in io.vertx.junit5.VertxExtensionTest
thisMustFail(VertxTestContext)  Time elapsed: 0.021 sec  <<< ERROR!
java.lang.RuntimeException: YOLO
	at io.vertx.junit5.VertxExtensionTest$EmbeddedWithARunner$DirectFailureTest.thisMustFail(VertxExtensionTest.java:143)

thisMustFail(Vertx, VertxTestContext)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: org.opentest4j.AssertionFailedError: 
	at io.vertx.junit5.VertxExtensionTest$EmbeddedWithARunner$FailureTest.lambda$thisMustFail$0(VertxExtensionTest.java:115)
	at io.vertx.junit5.VertxExtensionTest$EmbeddedWithARunner$FailureTest.thisMustFail(VertxExtensionTest.java:114)

Running io.vertx.junit5.AsyncBeforeEachTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.192 sec - in io.vertx.junit5.AsyncBeforeEachTest
Running io.vertx.junit5.CountingCheckpointTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.037 sec - in io.vertx.junit5.CountingCheckpointTest
Running io.vertx.junit5.IntegrationTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.161 sec - in io.vertx.junit5.IntegrationTest

Results :

Failed tests: 
  VertxExtensionTest$EmbeddedWithARunner$FailureTest.thisMustFail org.opentest4j.AssertionFailedError: 
Tests in error: 
  VertxExtensionTest$EmbeddedWithARunner$DirectFailureTest.thisMustFail:143 Runtime

Tests run: 33, Failures: 1, Errors: 1, Skipped: 0
```

Is that expected?

**Comment by jponge** [comment]
You are facing the surefire JDK 9 bug :-)

**Comment by jponge** [comment]
Try with JDK 8 when you run the project tests (not your tests)

**Comment by jponge** [comment]
Doing docs now, but the full life-cycle is supported.

**Comment by jponge** [comment]
Thanks @vietj @dadoonet for your contributions and feedbacks!

**Pull request merged**